docs/DeveloperGuide.md

```diff
@@ -2,14 +2,372 @@
 
 ## Acknowledgements
 
-{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
+{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the
+original source as well}
 
 ## Design & implementation
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### Main Class: `FAP`
 
+The `FAP` class serves as the central hub of the application, orchestrating the flow of execution and managing critical
+resources. It embodies key software design principles and showcases thoughtful architectural decisions.
+
+#### Design Overview
+
+- **Singleton Components:** `FAP` manages singleton components such as `moduleList` and `LOGGER`, ensuring they are
+  initialized once and accessible throughout the application's lifecycle.
+    - `ModuleList`: A collection that holds modules, initialized with a capacity of 10.
+    - `LOGGER`: Utilized for logging various levels of application events and errors.
+
+- **Separation of Concerns:** The class delegates specific responsibilities to specialized classes, adhering to the
+  principle of separation of concerns.
+    - Interaction with users is managed by the `Ui` class.
+    - Command parsing is delegated to the `Parser` class.
+    - Command execution is handled by implementations of the `Command` interface.
+
+- **Error Handling:** Demonstrates robust error handling strategies by catching and logging different exceptions, which
+  ensures graceful handling of unexpected situations.
+
+#### Implementation Details
+
+1. **Application Initialization and Entry Point:**
+
+   The `main` method, as the application's entry point, performs initial setups such as greeting the user and ensuring
+   critical components are initialized properly. It encapsulates high-level flow control and implements error handling
+   to manage assertion errors and unexpected exceptions.
+
+   ```java
+   public static void main(String[] args) {
+       try {
+           printGreeting();
+           assert moduleList != null : "moduleList should not be null";
+           runApplication();
+       } catch (AssertionError e) {
+           LOGGER.log(Level.SEVERE, "Assertion failed: " + e.getMessage(), e);
+           System.err.println("Critical assertion failure: " + e.getMessage());
+       } catch (Exception e) {
+           LOGGER.log(Level.SEVERE, "An unexpected error occurred: " + e.getMessage(), e);
+           System.err.println("An unexpected error occurred: " + e.getMessage());
+       }
+   }
+   ```
+   #### UML Diagram
+
+   ![FAP class diagram](diagrams/FAP.png)
```

Comment:
When defining classes in the class diagram, you can consider using '-' and + should be used to denote private and public attributes and methods respectively instead of using the default settings of plantUML.